src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/CompilationTagHelperResolver.cs

```diff
@@ -59,20 +58,22 @@ internal class CompilationTagHelperResolver(ITelemetryReporter? telemetryReporte
         static void ExecuteProviders(ITagHelperDescriptorProvider[] providers, TagHelperDescriptorProviderContext context, ITelemetryReporter? telemetryReporter)
         {
             using var _ = StopwatchPool.GetPooledObject(out var watch);
-            using var timingDictionary = new PooledDictionaryBuilder<string, object?>();
+            using var properties = new PooledArrayBuilder<Property>(providers.Length);
 
             foreach (var provider in providers)
             {
                 watch.Restart();
                 provider.Execute(context);
                 watch.Stop();
 
-                var propertyName = $"{provider.GetType().Name}.elapsedtimems";
-                Debug.Assert(!timingDictionary.ContainsKey(propertyName));
-                timingDictionary.Add(propertyName, watch.ElapsedMilliseconds);
+                if (telemetryReporter is not null)
+                {
+                    var propertyName = $"{provider.GetType().Name}.elapsedtimems";
```

Comment:
ℹ️ There's no need to allocate the property name and add the property if we're not going to send the telemetry.